src/util/alloc/allocator.rs

```diff
@@ -26,6 +36,10 @@ pub fn align_allocation<VM: VMBinding>(
     known_alignment: usize,
     fillalignmentgap: bool,
 ) -> Address {
+    if region.is_zero() {
```

Comment:
This was where one of the cases failed. Hmm I've deleted the test output files unfortunately. I will try to recreate it and see where the error occurred (from memory it was due to the `debug_assert!(!(fillalignmentgap && region.is_zero()));`).

Reply:
Right so it was an issue with `LargeObjectAllocator`:
```
RUST_BACKTRACE=1 RUST_LOG= MMTK_PLAN=SemiSpace ./build/linux-x86_64-normal-server-fastdebug/jdk/bin/java -XX:+UseThirdPartyHeap -Xms5M -Xmx5M OOMTest
thread '<unnamed>' panicked at 'assertion failed: !(fillalignmentgap && region.is_zero())', /home/kunals/git/mmtk-core/src/util/alloc/allocator.rs:49:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/panicking.rs:107:14
   2: core::panicking::panic
             at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/panicking.rs:48:5
   3: mmtk::util::alloc::allocator::align_allocation
             at /home/kunals/git/mmtk-core/src/util/alloc/allocator.rs:49:5
   4: <mmtk::util::alloc::large_object_allocator::LargeObjectAllocator<VM> as mmtk::util::alloc::allocator::Allocator<VM>>::alloc
             at /home/kunals/git/mmtk-core/src/util/alloc/large_object_allocator.rs:35:9
   5: <mmtk::plan::mutator_context::Mutator<VM> as mmtk::plan::mutator_context::MutatorContext<VM>>::alloc
             at /home/kunals/git/mmtk-core/src/plan/mutator_context.rs:83:9
   6: mmtk::memory_manager::alloc
             at /home/kunals/git/mmtk-core/src/memory_manager.rs:132:5
   7: alloc
             at /home/kunals/git/mmtk-openjdk/mmtk/src/api.rs:86:5
   8: _ZN18MMTkMutatorContext5allocEm9Allocator
             at ./../../openjdk/mmtkMutator.cpp:21
   9: _ZNK12MemAllocator21allocate_outside_tlabERNS_10AllocationE
             at ./src/hotspot/share/gc/shared/memAllocator.cpp:272
  10: _ZNK12MemAllocator12mem_allocateERNS_10AllocationE
             at ./src/hotspot/share/gc/shared/memAllocator.cpp:370
  11: _ZNK12MemAllocator8allocateEv
             at ./src/hotspot/share/gc/shared/memAllocator.cpp:377
  12: _ZN13CollectedHeap14array_allocateEP5KlassiibP6Thread
             at ./src/hotspot/share/gc/shared/collectedHeap.cpp:457
  13: _ZN13InstanceKlass17allocate_objArrayEiiP6Thread
             at ./src/hotspot/share/oops/instanceKlass.cpp:1223
  14: _ZN10oopFactory12new_objArrayEP5KlassiP6Thread
             at ./src/hotspot/share/memory/oopFactory.cpp:87
  15: _ZN18InterpreterRuntime9anewarrayEP10JavaThreadP12ConstantPoolii
             at ./src/hotspot/share/interpreter/interpreterRuntime.cpp:254
  16: <unknown>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
fatal runtime error: failed to initiate panic, error 5
```

Reply:
I think the problem is only for the LOS Allocator so I could potentially move the null check to it's allocation function instead of here. I will have to think more about if that change will work.

Reply:
Right -- it does seem like other allocators will use their cursor and the only place where this is could be an issue is the LOS allocator. I can't think of a case where cursor should ever be `null` (other than initialization of course) since we will never set the limit to be `null` (at least as far as I can see). I think the current `debug_assert!()` suffices as a check and I will move the null check to the LOS allocator.